include/cuda/memory_resource

```diff
@@ -0,0 +1,973 @@
+#include "std/version"
+#include "stream_view"
+
+#ifdef _LIBCUDACXX_EXT_RTTI_ENABLED
+#include <typeinfo>
+#endif
+
+#if _LIBCUDACXX_STD_VER > 14
+#if __has_include(<memory_resource>)
+#include <memory_resource>
+#define _LIBCUDACXX_STD_PMR_NS ::std::pmr
+#elif __has_include(<experimental/memory_resource>)
+#include <experimental/memory_resource>
+#define _LIBCUDACXX_STD_PMR_NS ::std::experimental::pmr
+#endif // __has_include(<experimental/memory_resource>)
+#endif // _LIBCUDACXX_STD_VER > 14
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+/*!
+ * \brief Groups the tag types denoting the kind of memory of an allocation.
+ *
+ * Memory allocation kind determines where memory can be accessed and the
+ * performance characteristics of accesses.
+ *
+ * This is not a closed set, the user code can define custom memory kinds.
+ */
+namespace memory_kind {
+  /*!
+   * \brief Ordinary host memory
+   */
+  struct host;
+
+  /*!
+   * \brief Device memory, as allocated by cudaMalloc.
+   */
+  struct device;
+
+  /*!
+   * \brief Device-accessible host memory.
+   */
+  struct pinned;
+
+  /*!
+   * \brief Virtual memory that is automatically migrated between the host and devices.
+   */
+  struct managed;
+};
+
+namespace detail {
+
+template <typename...>
+struct __type_pack {};
+
+namespace __fallback_typeid {
+
+template <class _Tp>
+struct  _LIBCUDACXX_TEMPLATE_VIS __unique_typeinfo { static constexpr int __id = 0; };
+template <class _Tp> constexpr int __unique_typeinfo<_Tp>::__id;
+
+template <class _Tp>
+inline _LIBCUDACXX_INLINE_VISIBILITY
+constexpr const void* __get_fallback_typeid() {
+    return &__unique_typeinfo<std::decay_t<_Tp>>::__id;
+}
+
+template <typename _Tp>
+const ::std::type_info *__get_typeid() {
+#ifdef _LIBCUDACXX_EXT_RTTI_ENABLED
+  return &typeid(_Tp);
+#else
+  return nullptr;
+#endif
+}
+
+inline bool __compare_type(const ::std::type_info *__ti1, const void *__fallback_ti1,
+                           const ::std::type_info *__ti2, const void *__fallback_ti2) {
+#ifdef _LIBCUDACXX_EXT_RTTI_ENABLED
+  if (__ti1 && __ti2 && *__ti1 == *__ti2)
+    return true;
+#endif
+  return __fallback_ti1 == __fallback_ti2;
+}
+
+template <typename _Tp>
+bool __is_type(const ::std::type_info *__ti1, const void *__fallback_ti1) {
+  return __compare_type(__ti1, __fallback_ti1, __get_typeid<_Tp>(), __get_fallback_typeid<_Tp>());
+}
+
+}  // namespace __fallback_typeid
+
+}  // namespace detail
+
+template <typename _ResourcePointer, typename... _Properties>
+class basic_resource_view;
+
+template <typename... _Properties>
+struct properties;
+
+
+namespace detail {
+  template <typename... _Properties>
+  struct __kind_from_properties_helper {
+    using type = void;
+  };
+
+  template <typename... _Properties>
+  using __kind_from_properties_helper_t = typename __kind_from_properties_helper<_Properties...>::type;
+
+  template <typename _Kind, typename... _Tail>
+  struct __kind_from_properties_helper<is_kind<_Kind>, _Tail...> {
```

Comment:
This code does not compile for me anymore after this latest change -- `is_kind` is not defined until later in the file.

```
deps/libcudacxx-src/include/cuda/memory_resource:129:40: error: 'is_kind' was not declared in this scope
  129 |   struct __kind_from_properties_helper<is_kind<_Kind>, _Tail...> {
      |                                        ^~~~~~~
_deps/libcudacxx-src/include/cuda/memory_resource:129:53: error: template argument 1 is invalid
  129 |   struct __kind_from_properties_helper<is_kind<_Kind>, _Tail...> {
      |                                                     ^
_deps/libcudacxx-src/include/cuda/memory_resource:149:3: error: wrong number of template arguments (2, should be 3)
  149 |   >;
```

Reply:
Fix for this appears to be to move the `struct is_kind` definition above this.